CLR/Libraries/Windows_Devices/windows_devices_native_Windows_Devices_Gpio_GpioPin.cpp

```diff
@@ -0,0 +1,234 @@
+{
+    TINYCLR_HEADER();
+
+    BOOL state;
+    CLR_RT_HeapBlock* pThis = stack.This(); FAULT_ON_NULL(pThis);
+
+    if (pThis[ FIELD__m_disposed ].NumericByRef().s1 != 0)
+    {
+        TINYCLR_SET_AND_LEAVE(CLR_E_OBJECT_DISPOSED);
+    }
+
+    pThis[ FIELD__m_lastOutputValue ].NumericByRef().s4 = stack.Arg1().NumericByRef().s4;
+    state = (stack.Arg1().NumericByRef().s4 != GpioPinValue_Low);
+
+    // If the current drive mode is set to output, write the value to the pin.
+    if (pThis[ FIELD__m_driveMode ].NumericByRefConst().s4 == GpioPinDriveMode_Output)
+    {
+        ::CPU_GPIO_SetPinState( pThis[ FIELD__m_pinNumber ].NumericByRefConst().u4, state );
+    }
+
+    TINYCLR_NOCLEANUP();
+}
+
+HRESULT Library_windows_devices_native_Windows_Devices_Gpio_GpioPin::Read___WindowsDevicesGpioGpioPinValue( CLR_RT_StackFrame& stack )
+{
+    TINYCLR_HEADER();
+
+    BOOL state;
+    CLR_RT_HeapBlock* pThis = stack.This(); FAULT_ON_NULL(pThis);
+
+    if (pThis[ FIELD__m_disposed ].NumericByRef().s1 != 0)
+    {
+        TINYCLR_SET_AND_LEAVE(CLR_E_OBJECT_DISPOSED);
+    }
+
+    state = ::CPU_GPIO_GetPinState( pThis[ FIELD__m_pinNumber ].NumericByRefConst().u4 );
+    stack.SetResult_I4( state ? GpioPinValue_High : GpioPinValue_Low );
+
+    TINYCLR_NOCLEANUP();
+}
+
+HRESULT Library_windows_devices_native_Windows_Devices_Gpio_GpioPin::InitNative___VOID__I4( CLR_RT_StackFrame& stack )
+{
+    TINYCLR_HEADER();
+
+    GPIO_PIN portId = GPIO_PIN_NONE;
+    BOOL pinAllocated = FALSE;
+    CLR_RT_HeapBlock* pThis = stack.This(); FAULT_ON_NULL(pThis);
+
+    portId = stack.Arg1().NumericByRef().u4;
+
+    // Ensure the pin exists.
+    if (::CPU_GPIO_Attributes( portId ) == 0)
+    {
+        TINYCLR_SET_AND_LEAVE(CLR_E_INVALID_PARAMETER);
+    }
+
+    // Reserve the pin. If this fails, it's already in use.
+    pinAllocated = ::CPU_GPIO_ReservePin( portId, TRUE );
+    if (!pinAllocated)
+    {
+        TINYCLR_SET_AND_LEAVE(CLR_E_PIN_UNAVAILABLE);
+    }
+
+    pThis[ FIELD__m_pinNumber ].NumericByRef().u4 = portId;
+
+    // Initialize the default drive mode.
+    if (!::CPU_GPIO_EnableInputPin2( portId, false, NULL, NULL, GPIO_INT_NONE, RESISTOR_DISABLED ))
+    {
+        TINYCLR_SET_AND_LEAVE(CLR_E_INVALID_OPERATION);
+    }
+
+    TINYCLR_CLEANUP();
+    if (FAILED(hr))
+    {
+        if (pinAllocated)
+        {
+            ::CPU_GPIO_ReservePin( portId, FALSE );
+        }
+    }
+    TINYCLR_CLEANUP_END();
+}
+
+HRESULT Library_windows_devices_native_Windows_Devices_Gpio_GpioPin::DisposeNative___VOID( CLR_RT_StackFrame& stack )
+{
+    TINYCLR_HEADER();
+
+    GPIO_PIN portId;
+    CLR_RT_HeapBlock* pThis = stack.This(); FAULT_ON_NULL(pThis);
+
+    portId = pThis[ FIELD__m_pinNumber ].NumericByRefConst().u4;
+
+    // Set pin to input state so the pin is in "weak" state and does not drain power. Then release the pin.
+    ::CPU_GPIO_EnableInputPin2( portId, false, NULL, NULL, GPIO_INT_NONE, RESISTOR_PULLUP );

```

Comment:
When we update the HAL to support per pin de-bounce, let's also update it to include an explicit DisablePin() function that way it is semantically clearer on what this is doing and allows the HAL implementation to do whatever is appropriate as the default here using Input mode with a pull-up, may not actually be the best way to do that on a given piece of hardware.